File: src/chain.js
/* Chain projections. The event log is the single source of truth; everything
   here derives from it and holds no state of its own.

   Every entry's digest is SHA-256 over its full content plus its
   predecessor's digest, and an enrolled actor's entry is signed Ed25519
   over that digest. Identity custody lives outside the record. */

import { sha256Hex, signText } from "./crypto.js";

export const GENESIS = "0".repeat(64);
/* Canonical entry digest (format v3). Every content field is committed, so
   altering any of them after the fact breaks the chain at exactly that
   entry, including the anchored artifact: its file hash, name, size, and
   (when the audio is uncompressed) its PCM hash all commit, so neither
   the file, its name, nor the samples themselves can be swapped. v3 also
   commits the derivation link (`derivedFrom`, an earlier entry's seq) and
   the reserved signature-scheme field (`alg`; absent means Ed25519).
   The reader recomputes this same function over a package it is handed.
   v2 never shipped; clean break,
   same as v1→v2. */
export const entryHash = (prev, i, e) => {
  const actor = e.actor || { id: null, name: "unresolved" };
  const artifact = e.artifact
    ? `${e.artifact.sha256}:${e.artifact.name}:${e.artifact.size}:${e.artifact.pcm_sha256 ?? ""}`
    : "";
  return sha256Hex(
    `${prev}|${i}|${e.action}|${e.lane}|${actor.id}|${actor.name}|${e.m}|${e.room}|${e.note ?? ""}|${artifact}|${e.derivedFrom ?? ""}|${e.alg ?? ""}`,
  );
};

/* Session-relative clock from a 14:00 downbeat. Skeleton-stage: positions
   inside one session. Real wall-clock timestamps arrive with persistence. */
export const hhmm = (m) =>
  `${String(14 + Math.floor(m / 60)).padStart(2, "0")}:${String(m % 60).padStart(2, "0")}`;

/* Adds seq, t, sealed, prev, hash, and — when the actor has a signing key —
   an Ed25519 signature over the entry's digest. Derives defensively: events
   are sorted by session time first, and an actor without an identifier is a
   recorded gap, never an error. Sequential by construction: each digest
   commits to the one before it. */
export const buildEvents = async (rawEvents, signerFor = null) => {
  const ordered = [...rawEvents].sort((a, b) => (a.m ?? 0) - (b.m ?? 0));
  let prev = GENESIS;
  const out = [];
  for (let i = 0; i < ordered.length; i++) {
    const e = ordered[i];
    const actor = e.actor || { id: null, name: "unresolved" };
    const sealed = Boolean(actor.id);
    const hash = await entryHash(prev, i, { ...e, actor });
    let sig;
    if (sealed && signerFor) {
      const key = signerFor(actor.id);
      if (key) sig = await signText(key, hash);
    }
    out.push({ ...e, actor, seq: i + 1, t: hhmm(e.m ?? 0), sealed, prev, hash, ...(sig ? { sig } : {}) });
    prev = hash;
  }
  return out;
};

export const verdictOf = (events, laneId) => {
  const own = events.filter((e) => e.lane === laneId);
  if (!own.length) return { key: "pending", label: "no events" };
  if (!own[0].sealed) return { key: "unverified", label: "origin unverified" };
  const gap = own.find((e) => !e.sealed);
  if (gap) return { key: "broken", label: "chain broken", gap };
  return { key: "intact", label: "chain intact" };
};

File: conformance/go/chain.go
package conformance

// The digest and the session clock, FORMAT.md 4.1 and 4.3.

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math"
	"strings"
)

const Genesis = "0000000000000000000000000000000000000000000000000000000000000000"
const Format = "sealedrecord/package.v3"

func sha256Hex(b []byte) string {
	s := sha256.Sum256(b)
	return hex.EncodeToString(s[:])
}

func actorOf(e map[string]any) map[string]any {
	if a, ok := e["actor"].(map[string]any); ok && a != nil {
		return a
	}
	return map[string]any{"id": nil, "name": "unresolved"}
}

// optional is the primed component of 4.1: the value when present and not
// null, else the empty string.
func optional(e map[string]any, k string) string {
	if v, ok := e[k]; ok && v != nil {
		return stringify(v, true)
	}
	return ""
}

func artifactComponent(e map[string]any) string {
	a, ok := e["artifact"].(map[string]any)
	if !ok || a == nil {
		return ""
	}
	return str(a, "sha256") + ":" + str(a, "name") + ":" + str(a, "size") + ":" + optional(a, "pcm_sha256")
}

// EntryHash is preimage_i of 4.1, hashed.
func EntryHash(prev string, i int, e map[string]any) string {
	actor := actorOf(e)
	pre := strings.Join([]string{
		prev, fmt.Sprint(i), str(e, "action"), str(e, "lane"), str(actor, "id"), str(actor, "name"),
		str(e, "m"), str(e, "room"), optional(e, "note"), artifactComponent(e), optional(e, "derivedFrom"), optional(e, "alg"),
	}, "|")
	return sha256Hex([]byte(pre))
}

func pad2(s string) string {
	if len(s) < 2 {
		return strings.Repeat("0", 2-len(s)) + s
	}
	return s
}

// HHMM is 4.3 for a numeric m. A non-numeric m stringifies as JavaScript
// would arrive at NaN, which can never equal a recorded t.
func HHMM(mv any, present bool) string {
	m, ok := number(mv)
	if !ok {
		if !present || mv == nil {
			m = 0 // JavaScript: `m ?? 0` in the producer; the reader uses m as given, null/60 = 0
		} else {
			return "NaN"
		}
	}
	h := 14 + math.Floor(m/60)
	r := math.Mod(m, 60)
	return pad2(jsFloat(h)) + ":" + pad2(jsFloat(r))
}

func jsFloat(f float64) string {
	if f == math.Trunc(f) {
		return fmt.Sprintf("%.0f", f)
	}
	return fmt.Sprintf("%g", f)
}
